DataFormats/HcalRecHit/interface/HFQIE10Info.h

```diff
@@ -47,7 +83,22 @@ class HFQIE10Info
     // we are going to check all time slices. Otherwise only
     // the "sample of interest" time slice is checked.
     //
-    bool isDataframeOK(bool checkAllTimeSlices = false) const;
+    constexpr bool isDataframeOK(bool checkAllTimeSlices = false) const {
+        bool hardwareOK = true;
+        if (soi_ >= nRaw_ || checkAllTimeSlices)
+            for (unsigned i=0; i<nRaw_ && hardwareOK; ++i) {
+                const QIE10DataFrame::Sample s(
+                    (edm::DataFrame::data_type)(raw_[i] & 0xffff), 
+                    (edm::DataFrame::data_type)((raw_[i]>>16) & 0xffff));
```

Comment:
.cc had ```const QIE10DataFrame::Sample s(raw_[i]);``` which was using ```explicit Sample(wide_type wide)```.
Was there a reason that this constructor from wide_type could not be made constexpr?
It looks wrong to have to reimplement the method ```Sample(wide_type wide)``` with this in situ unpacking alternative

Reply:
`reinterpret_cast` seems not to be allowed in constexpr functions

Reply:
This could be a problem... One of the rechit constructor arguments is the array of raw data generated by cycling over the dataframe samples and calling the "wideRaw()" method of QIE10DataFrame::Sample. The QIE10DataFrame::Sample(wide_type wide) does the right thing no matter what the endianness is. The new code, however, seems to be endian-dependent.

To me, it looks safer to put the C-style cast (replacing reinterpret_cast) inside QIE10DataFrame::Sample(wide_type wide) and QIE10DataFrame::wideRaw() instead of here.

Reply:
@igv4321 i think i incorrectly sent my previous comment... so reposting again:

are you sure here that `QIE10DataFrame:: Sample`  is endianness independent???

from here
https://github.com/cms-sw/cmssw/blob/master/DataFormats/HcalDigi/interface/QIE10DataFrame.h#L31
```
    explicit Sample(const wide_type wide) 
      : word1_{0}, word2_{0} {
      static_assert(sizeof(wide) == 2*sizeof(word1_),
                "The wide input type must be able to contain two words");
      const edm::DataFrame::data_type* ptr =
        reinterpret_cast<const edm::DataFrame::data_type*>(&wide);
      word1_ = ptr[0];
      word2_ = ptr[1];
    }
```

given 32 bit logical value
- word1 is the lowest 16 bits of that
- word2 is the highest 16 bits of that

for 
- little endian, `ptr[0]` will give you the lowest 2 bytes (`word1`) and `ptr[1]` will give us highest 2 bytes (`word2`)
- big endian, vice versa

unless the bytes are flipped by reinterpret_cast...

am i missing incorrect here??? cause i think that current impl is actually little endian specific...

Reply:
What has to happen is that "wideRaw" method of QIE10DataFrame::Sample and "QIE10DataFrame::Sample(const wide_type wide) need to be in sync. They have to pack and unpack 16-bit values in a consistent manner. If one does it through a cast, the other should do it in the same way.

Reply:
This needs to be resolved for the PR to proceed.

Reply:
the previous comment did not address my concern/comment that I've made...  i did not touch the QIE10DataFrame impl itself... it is as it was...

repeating again,  i did not want to change the current implementation, but simply added a bit of stuff to be endianness agnostic in HFQIE10Info.h (bit operations applied should be endianness agnostic) member function, which does not change the state of the class, but simply uses the ctor that fits the needs here... 

i will take any decision, but i need to know what you want me to modify??? I think i laid out my opinion on the changes made in my previous comment.... 
Please let me know!

Reply:
Yes, but you are now remaking QIE10DataFrame::Sample from a wide_type using a constructor which was not intended for that purpose and in a manner not necessarily consistent with the code that makes that wide_type to begin with. QIE10DataFrame::Sample(wide_type wide) did it consistently.

I suggest reverting to the previous version of "isDataframeOK"method and keeping the code which packs and unpacks wide_type inside QIE10DataFrame::Sample. If it requires changing reinterpret_cast into the C-style cast, so be it.

Reply:
I would not say that it is not consistent, in the current impl little endian is assumed and bytes 0,1 go to word1 and 2,3 bytes go to word2... i do exactly the same... in principle

in c++, c style cast is specified here http://en.cppreference.com/w/cpp/language/explicit_cast

and when trying to replace reinterpret_cast with c-style cast, it will not compile...

```
error: constexpr function never produces a constant expression [-Winvalid-constexpr]
    constexpr QIE10DataFrame::Sample::wide_type wideRaw() const {
                                                ^
/data/user/vkhriste/cmssw_releases/CMSSW_10_2_X_2018-06-18-2300/src/DataFormats/HcalDigi/interface/QIE10DataFrame.h:59:9: note: cast that performs the conversions of a reinterpret_cast is not allowed in a constant
      expression
        (edm::DataFrame::data_type*)&result;
```

another suggestion is to remove reinterpret_cast in both wideRaw and ctor and put bit shifting instead...

or, in the end, i can leave isDataFramOK on the host side as is and just create a device side analog...

further suggestions??? 
thanks!

VK

Reply:
another suggestion,

I can skip this particular function for now.... not mark it as constexpr as it's not a show stopper for further devs...

VK

Reply:
I've undone the constexpr for this part...

Reply:
> I can skip this particular function for now.... not mark it as constexpr as it's not a show stopper for further devs...

works for me.
Thanks